streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -1767,6 +1780,15 @@ class ZStream[-R, +E, +A] private[stream] (private[stream] val structure: ZStrea
       } yield pull
     }
 
+  /**
+   * Interrupts the evaluation of this stream when the provided promise resolves. This
+   * combinator will also interrupt any in-progress element being pulled from upstream.
+   *
+   * If the promise completes with a failure, the stream will emit that failure.
+   */
+  final def interruptWhen[E1 >: E](p: Promise[E1, _]): ZStream[R, E1, A] =
```

Comment:
Could you change the signature while you're here?
```suggestion
  final def interruptWhen[E1 >: E](p: Promise[E1, Any]): ZStream[R, E1, A] =
```

Reply:
Not sure this would have the desired outcome – `Promise` is invariant, so you'd have to exactly pass a `Promise[E1, Any]`. The existential is unfortunate here, but it doesn't appear to be avoidable unless `Promise` is made covariant.

The `IO` variants could certainly take `IO[E, Any]`, but now that I think about it, that's basically tantamount to a value discard, which is usually considered a wart. So I changed them to take `IO[E, Unit]` (it's easy to explicitly discard the value by just calling `.unit` on the argument)